app/services/attachments/concatenation_service.rb

```diff
@@ -0,0 +1,153 @@
+
+    attr_accessor :attachable, :attachments, :concatenation_params
+
+    def initialize(user = nil, attachable = nil, params = {})
+      super(user, params)
+      @attachable = attachable
+
+      # single-end params: { attachment_ids = [], basename: basefilename OPTIONAL,
+      #                      delete_originals: true OPTIONAL
+      #                      }
+      # paired-end params: { attachment_ids = [[],[]], basename: basefilename OPTIONAL,
+      #                      delete_originals: true OPTIONAL
+      #                      }
+      @concatenation_params = params
+    end
+
+    def execute # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+      # authorize if user can update sample
+      authorize! attachable.project, to: :update_sample? if attachable.instance_of?(Sample)
+
+      attachment_ids = concatenation_params[:attachment_ids]
+
+      unless attachment_ids.all? { |i| i.is_a?(Integer) }
+        # if multi-dimensional array of ids
+        attachment_ids = attachment_ids.first + attachment_ids.last
+      end
+
+      attachments = attachable.attachments.where(id: attachment_ids, attachable:).order(:id)
+
+      if attachments.length != attachment_ids.length
+        raise AttachmentConcatenationError,
+              I18n.t('services.attachments.concatenation.incorrect_attachable')
+      end
+
+      if attachments.length.positive?
+        is_paired_end = attachments.first.metadata.key?('type')
+        is_single_end = !attachments.first.metadata.key?('type')
+
+        validate_same_types(attachments, is_paired_end, is_single_end)
+
+        if is_paired_end
+          forward_reads = []
+          reverse_reads = []
+          attachments.each do |attachment|
+            if attachment.metadata['direction'] == 'forward'
+              forward_reads << attachment
+            elsif attachment.metadata['direction'] == 'reverse'
+              reverse_reads << attachment
+            end
+          end
+          concatenate_paired_end_reads(forward_reads, reverse_reads)
+        elsif is_single_end
+          concatenate_single_end_reads(attachments)
+        end
+
+        # if option is selected then destroy the original files
+        attachments.each(&:destroy!) if concatenation_params[:delete_originals]
+      end
+    rescue Attachments::ConcatenationService::AttachmentConcatenationError => e
+      attachable.errors.add(:base, e.message)
+      attachable
+    end
+
+    private
+
+    # Validates attachments to ensure they are of the correct type.
+    # Files can only be concatenated as follows. No mixing of types:
+    # Paired-end -> Paired-end
+    # Single-end -> Single-end
+    def validate_same_types(attachments, is_paired_end, is_single_end) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+      valid_type = true
+
+      attachments.each do |attachment|
```

Comment:
**nitpick (blocking):** Reading this code `valid_type` could go from true to false to true. E.g. first attachment set's it to false and second sets it true. Also instead of checking for presence of key we should check the value. I also wonder if we should separate validation of extensions into a separate method to reduce rubocop disables.

Reply:
Updated this so valid_type is no longer used. Instead we continue going through the attachments unless one of them doesn't match the `if` at which point an error will be raised in 349d6987f0f58fb6c2481ed054fe66446e6a6090